chsdi/views/catalog.py

```diff
@@ -2,13 +2,83 @@
 
 from pyramid.view import view_config
 from pyramid.httpexceptions import HTTPNotFound
+import networkx as nx
+from itertools import chain
+from pyramid.httpexceptions import HTTPInternalServerError
 
 from chsdi.models.bod import Catalog
 from chsdi.lib.validation import MapNameValidation
 from chsdi.lib.sqlalchemy_customs import remove_accents
 from chsdi.lib.filters import filter_by_geodata_staging
 
 
+# https://github.com/networkx/networkx/blob/master/networkx/readwrite/json_graph/tree.py#L16
+def tree_data(G, root, attrs, meta):
+    if G.number_of_nodes() != G.number_of_edges() + 1:
+        raise TypeError("G is not a tree.")
+    if not G.is_directed():
+        raise TypeError("G is not directed.")
+
+    id_ = attrs['id']
+    children = attrs['children']
+    if id_ == children:
+        raise nx.NetworkXError('Attribute names are not unique.')
+
+    def add_children(n, G):
+        nbrs = G[n]
+        if len(nbrs) == 0:
+            return []
+        children_ = []
+        nbrs = sorted(nbrs)
+        for child in nbrs:
+            d = dict(chain(G.node[child].items(), [(id_, child)]))
+            d['category'] = meta[d['id']]['category']
+            d['staging'] = meta[d['id']]['staging']
+            d['label'] = meta[d['id']]['label']
+            # only for sorting according to 'orderKey'
+            if ('orderKey' in meta[d['id']]):
+                d['orderKey'] = meta[d['id']]['orderKey']
+            if (d['category'] == 'layer'):
+                d['layerBodId'] = meta[d['id']]['layerBodId']
+            else:
+                d['selectedOpen'] = meta[d['id']]['selectedOpen']
+            c = add_children(child, G)
+            if c:
+                d[children] = c
+            children_.append(d)
+            if ('orderKey' in d):
+                order_key = lambda x: x['orderKey']
+                children_ = sorted(children_, key=order_key)
+        for d in children_:
+            d.pop('orderKey', None)
+        return children_
+
+    data = dict(chain(G.node[root].items(), [(id_, root)]))
+    data[children] = add_children(root, G)
+    return data
+
+
+def initialize_graph(G, rows, lang, stage):
+    meta = {}
+    for r in rows:
+        node_parent_id = getattr(r, 'parentId')
+        node_id = getattr(r, 'id')
+        if node_id not in meta:
+            meta[node_id] = r.to_dict(lang)
+        G.add_edge(node_parent_id, node_id)
+        if not node_parent_id:
+            root_id = node_id
+    return G, meta, root_id
+
+
+def create_digraph(rows, lang, stage):
+    G = nx.DiGraph()
+    graph, meta, root_id = initialize_graph(G, rows, lang, stage)
+    edge_list = list(nx.bfs_edges(graph, root_id))
+    G = nx.DiGraph(edge_list)
+    return G, meta, root_id
+
+
```

Comment:
Ok. I think it's worth it to mention it in the code (that code is not 1-1, but inspired...)